lib/HLSL/DxilValidation.cpp

```diff
@@ -3988,6 +3988,114 @@ static void ValidateMetadata(ValidationContext &ValCtx) {
   }
 
   const hlsl::ShaderModel *SM = ValCtx.DxilMod.GetShaderModel();
+  // validate that any wavesize tags don't appear outside their expected shader
+  // models. Validate only 1 tag exists per entry point.
+  // Don't do this validation if the shader is non-compute
+  if (SM->IsCS() || SM->IsLib()) {
+
+    NamedMDNode *TA = pModule->getNamedMetadata("dx.entryPoints");
+    if (TA != nullptr) {
+      for (unsigned i = 0, end = TA->getNumOperands(); i < end; ++i) {
+        MDTuple *TANode = dyn_cast<MDTuple>(TA->getOperand(i));
+        if (TANode->getNumOperands() < 5) {
+          ValCtx.EmitMetaError(TANode, ValidationRule::MetaWellFormed);
+          return;
+        }
+        // get access to the digit that represents the metadata number that
+        // would store entry properties
+        const llvm::MDOperand &mOp =
+            TANode->getOperand(TANode->getNumOperands() - 1);
+        // the final operand to the entry points tuple should be a constant int.
+        if (mOp == nullptr ||
+            (mOp.get())->getMetadataID() != Metadata::MDTupleKind) {
+          continue;
+        }
+
+        // get access to the node that stores entry properties
+        MDTuple *TANode2 =
+            dyn_cast<MDTuple>(TANode->getOperand(TANode->getNumOperands() - 1));
+        // find any incompatible tags inside the entry properties
+        // increment j by 2 to only analyze tags, not values
+        bool foundTag = false;
+        for (unsigned j = 0, end2 = TANode2->getNumOperands(); j < end2;
+             j += 2) {
+          // if this tag is already present in the current entry point, emit an
+          // error
```

Comment:
If you do break this code out into a separate function, you might take a pass at awkward comment breaks like this since they should be less deeply embedded in that case.